lib/puppet/http/client.rb

```diff
@@ -177,8 +177,8 @@ def head(url, headers: {}, params: {}, options: {})
   #
   # @return [String] the body of the request response
   #
-  def put(url, headers: {}, params: {}, options: {})
-    body = options.fetch(:body) { |_| raise ArgumentError, "'put' requires a 'body' option" }
+  def put(url, body, headers: {}, params: {}, options: {})
+    raise ArgumentError, "'put' requires a 'body' argument" unless body
```

Comment:
```
irb(main):006:1* def put(url, body, headers: {}, params: {}, options: {})
irb(main):007:1*   puts body
irb(main):008:0> end
=> :put
irb(main):009:0> put('url.example.com', 'asd',  headers: {stuff: 'asdf'})
asd
=> nil
irb(main):010:0> put('url.example.com', headers: {stuff: 'asdf'})
{:headers=>{:stuff=>"asdf"}}
=> nil
```

Reply:
For me ruby rejects the call, but maybe it is ruby version dependent?
```
irb(main):007:0> def put(url, body, headers: {}, params: {}, options: {})
irb(main):008:1> puts body
irb(main):009:1> end
=> :put
irb(main):010:0> put('url')
ArgumentError: wrong number of arguments (given 1, expected 2)
	from (irb):7:in `put'
	from (irb):10
	from /usr/local/opt/rbenv/versions/2.3.8/bin/irb:11:in `<top (required)>'
...
irb(main):011:0> RUBY_VERSION
=> "2.3.8"
```

Reply:
oh sorry, it's the `headers` argument. Ruby interprets that to mean the second positional argument... ugh

Reply:
It looks like if you include keyword arguments, but not a body, it'll use the keyword arguments for the body?

```
irb(main):001:1* def put(url, body, headers: {}, params: {}, options: {})
irb(main):002:1*   puts body
irb(main):003:0> end
=> :put
irb(main):004:0> put('url')
Traceback (most recent call last):
       16: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-1.17.2/exe/bundle:30:in `block in <top (required)>'
       15: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli.rb:18:in `start'
       14: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
       13: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli.rb:27:in `dispatch'
       12: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
       11: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
       10: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        9: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli.rb:463:in `exec'
        8: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli/exec.rb:28:in `run'
        7: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli/exec.rb:74:in `kernel_load'
        6: from /Users/melissa/.rbenv/versions/2.6.3/lib/ruby/2.6.0/bundler/cli/exec.rb:74:in `load'
        5: from /Users/melissa/workdir/puppet/.bundle/gems/bin/irb:23:in `<top (required)>'
        4: from /Users/melissa/workdir/puppet/.bundle/gems/bin/irb:23:in `load'
        3: from /Users/melissa/workdir/puppet/.bundle/gems/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        2: from (irb):4
        1: from (irb):1:in `put'
ArgumentError (wrong number of arguments (given 1, expected 2))
irb(main):005:0> put('url', headers: {'accept' => 'stuff'})
{:headers=>{"accept"=>"stuff"}}
=> nil
irb(main):006:0> RUBY_VERSION
=> "2.6.3"
```

Reply:
ha! Jinx!

Yeah, that seems like less than ideal behavior for ruby... it makes me wonder if we should also add a check to make sure url is set?

Reply:
To eliminate ambiguity we could make `url` and `body` keywords, but then you end up with `put(url: URI("https://..."))` which feels overly key-wordy

Reply:
I agree it feels overly key-wordy... I do feel like this is a ruby bug, so I'm fine leaving these as positional arguments. It makes me wonder if the test and raise if `body` is missing is even necessary?

Reply:
I think it'll be common for someone to do `client.post(url)` so I'm in favor of an explicit check for nil body. But we could make the check more restrictive to require it to be an instance of a string (which implies it's non-nil). Something like `'post' requires a string 'body' argument`? So that'd guard  against `client.post(url, headers: { .. })` and avoid the cryptic `NoMethodError (undefined method 'bytesize' for {:headers=>{}}:Hash)`.

Reply:
yeah, let's do that. I think it's a good idea that'll hopefully save us some headache in the future

Reply:
updated